konnectivity-client/pkg/client/client.go

```diff
@@ -123,8 +130,10 @@ func (t *grpcTunnel) serve(c clientConn) {
 					connid: resp.ConnectID,
 				}
 				select {
-				case ch <- result:
-				default:
+				// try to send to the result channel
+				case pendingDial.resultCh <- result:
```

Comment:
So this can leak goroutines. cancelCh is only closed if DialContext exits. So if stream.Send below blocks (it is a blocking call) then the read will never be called and DialContext will not exit.

Reply:
is it possible for serve to get a DIAL_RSP and for the `Send` that initiated the dial to still be blocked?

Reply:
sure... but if it is blocked and hasn't sent the PacketType_DIAL_REQ message yet, can't we be 100% sure that the `serve` loop won't be handling a PacketType_DIAL_RESP response for it yet?

Reply:
This still handles the race condition in the two scenarios described below - a second DIAL_RSP and DialContext() returning early because of timeout or context cancellation

I'm not at all familiar with the `Send()` behavior, but it would really surprise me if it were possible for `Send()` to remain blocked after the DIAL_REQ message has actually been sent and received and acted on... I can't tell if you're saying that's possible or not

Reply:
> put another way, does Send() block after sending a DIAL_REQ message completely enough that the stream receiver got it and is responding with a DIAL_RESP message for serve to handle?

Good point. If the DIAL_REQ send blocks, your right we won't get a DIAL_RESP message. So we'll only leak one side of the connection.

Reply:
> I assume by block you mean "it is a blocking call that gets stucked", instead of "it is a blocking call that might take a few ms to execute" (as most blocking send can do, when they don't sync with the receiver).

Correct. In this instance I'm not worried about it blocking for a few ms or even seconds. My concern is we know from painful experience that on a quota imbalance that the stream.Send() call can block forever.

Reply:
@cheftako oohh, that quota thing... good point. I'm not familiar with that, nor I have observed it myself. But yeah, that is something I missed, thanks

Reply:
I agree flowcontrol / quota could make `Send()` block for all sorts of reasons *before* completing sending the request, but it doesn't block *after*, right?